### ISSUE TITLE
Suggestion for `needless_borrow` can cause infinite recursion

### ISSUE THREAD

**Issue by u32i64** [rust-lang/rust-clippy#9095]
### Summary

According to https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow, this lint is `MachineApplicable`, but it appears that it can cause infinite recursion when applied.

### Lint Name

needless_borrow

### Reproducer

The following code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8d441748284503b19c331f49719b81fc)):

```rust
struct S(u32);

impl From<S> for u32 {
    fn from(s: S) -> Self {
        (&s).into()
    }
}

impl From<&S> for u32 {
    fn from(s: &S) -> Self {
        s.0
    }
}

fn main() {
    let s = S(1);
    let ref_s = &s;
    
    let a: u32 = ref_s.into();
    println!("{a}");
    
    let b: u32 = s.into();
    println!("{b}");
}
```

produces this warning:

```
    Checking playground v0.0.1 (/playground)
warning: this expression borrows a value the compiler would automatically borrow
 --> src/main.rs:5:9
  |
5 |         (&s).into()
  |         ^^^^ help: change this to: `s`
  |
  = note: `#[warn(clippy::needless_borrow)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.86s
```

Note that the code runs as expected in this case:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 1.16s
     Running `target/debug/playground`
1
1
```

But if I apply the suggestion, i.e.

```rust
impl From<S> for u32 {
    fn from(s: S) -> Self {
        s.into()
    }
}
```

Then there is no warning, but:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 1.61s
     Running `target/debug/playground`
1
thread 'main' has overflowed its stack
fatal runtime error: stack overflow
timeout: the monitored command dumped core
/playground/tools/entrypoint.sh: line 11:    10 Aborted                 timeout --signal=KILL ${timeout} "$@"
```

### Version

```text
rustc 1.64.0-nightly (46b8c23f3 2022-07-01)
binary: rustc
commit-hash: 46b8c23f3eb5e4d0e0aa27eb3f20d5b8fc3ed51f
commit-date: 2022-07-01
host: aarch64-apple-darwin
release: 1.64.0-nightly
LLVM version: 14.0.6
```

**Comment by u32i64** [comment]
I forgot to mention, it looks like this may be a regression. For example, on `nightly-x86_64-unknown-linux-gnu` in GitHub Actions:

- `rustc 1.64.0-nightly (7425fb293 2022-06-30)` didn't produce this clippy warning ([log](https://github.com/u32i64/signalupdates-bot/runs/7157550084?check_suite_focus=true)).

- `rustc 1.64.0-nightly (46b8c23f3 2022-07-01)` does produce the warning ([log](https://github.com/u32i64/signalupdates-bot/runs/7161655967?check_suite_focus=true#step:5:204)).
